File: study/ts/script/manager/HeroVoiceManager.ts
import { G_AudioManager, G_UserData, G_SceneManager } from "../init";
import { HeroAudioHelper } from "../utils/HeroAudioHelper";

export default class HeroVoiceManager {
    _curHeroVoice: any;
    _curHeroId: number;
    _isInMainMenu: boolean;
    _mainMenuScheduler: any;
    constructor() {
        this._curHeroVoice = null;
        this._curHeroId = 0;
        this._isInMainMenu = false;
        this._mainMenuScheduler = null;
    }
    clear() {
        this._curHeroVoice = null;
        this._curHeroId = 0;
    }
    setIsInMainMenu(isIn) {
        this._isInMainMenu = isIn;
    }
    playVoiceWithHeroId(heroId, must?) {
        var res = HeroAudioHelper.getVoiceRes(heroId);
        var play = function () {
            if (this._curHeroVoice) {
                G_AudioManager.stopSound(this._curHeroVoice);
            }
            var voice = null;
            if (res) {
                voice = G_AudioManager.playSound(res);
            }
            this._curHeroVoice = voice;
            this._curHeroId = heroId;
        }.bind(this);
        if (must) {
            play();
        } else if (heroId != this._curHeroId) {
            play();
        }
    }
    playCurRoleVoice() {
        var roleId = G_UserData.getHero().getRoleBaseId();
        this.playVoiceWithHeroId(roleId);
    }
    playRoleVoiceWithSex(sex) {
        var heroId = 1;
        if (sex == 1) {
            heroId = 1;
        } else {
            heroId = 11;
        }
        this.playVoiceWithHeroId(heroId);
    }
    createScheduler(callbcak, interval) {
    }
    stopScheduler(scheduleHandler) {
        if (scheduleHandler) {
            G_SceneManager.unschedule(scheduleHandler);
        }
        scheduleHandler = null;
    }
    startPlayMainMenuVoice() {
        var heroIds = G_UserData.getTeam().getHeroBaseIdsInBattle();
        function getInterval() {
            return Math.randInt(15, 20);
        }
        var playHeroVoice = function () {
            if (this._isInMainMenu) {
                var index = Math.randInt(0, heroIds.length - 1);
                var heroId = heroIds[index];
                this.playVoiceWithHeroId(heroId);
            }
        }.bind(this);
        this._mainMenuScheduler = function () {
            playHeroVoice();
            if (this._isInMainMenu) {
                playLoop();
            }
        }.bind(this);
        var playLoop = function () {
            var interval = getInterval();
            if(this._mainMenuScheduler)
            G_SceneManager.scheduleOnce(this._mainMenuScheduler, interval);
        }.bind(this);
        this.playCurRoleVoice();
        playLoop();
    }
    stopPlayMainMenuVoice() {
        if (this._mainMenuScheduler) {
            G_SceneManager.unschedule(this._mainMenuScheduler);
            this._mainMenuScheduler = null;
        }
    }
}

File: study/lua/src/app/manager/HeroVoiceManager.lua
--武将语音管理
local AudioManager = require("app.manager.AudioManager")
local HeroVoiceManager = class("HeroVoiceManager")
local HeroAudioHelper = require("app.utils.HeroAudioHelper")
local SchedulerHelper = require("app.utils.SchedulerHelper")

function HeroVoiceManager:ctor()
	self._curHeroVoice = nil
	self._curHeroId = 0
	self._isInMainMenu = false
	self._mainMenuScheduler = nil
end

function HeroVoiceManager:clear()
	self._curHeroVoice = nil
	self._curHeroId = 0
end

function HeroVoiceManager:setIsInMainMenu(isIn)
	self._isInMainMenu = isIn
end

function HeroVoiceManager:playVoiceWithHeroId(heroId, must)
	local res = HeroAudioHelper.getVoiceRes(heroId)

	local function play()
		if self._curHeroVoice then
			G_AudioManager:stopSound(self._curHeroVoice)
		end
		local voice = nil
		if res then
			voice = G_AudioManager:playSound(res)
		end
		
		self._curHeroVoice = voice
		self._curHeroId = heroId
	end

	if must then --一定播
		play()
	elseif heroId ~= self._curHeroId then
		play()
	end
end

--播放当前主角语音
function HeroVoiceManager:playCurRoleVoice()
	local roleId = G_UserData:getHero():getRoleBaseId()
	self:playVoiceWithHeroId(roleId)
end

--播放男\女主角语音
function HeroVoiceManager:playRoleVoiceWithSex(sex)
	local heroId = 1
	if sex == 1 then
		heroId = 1 --男
	else
		heroId = 11 --女
	end

	self:playVoiceWithHeroId(heroId)
end

function HeroVoiceManager:createScheduler(callbcak, interval)
	
end

function HeroVoiceManager:stopScheduler(scheduleHandler)
	if scheduleHandler then
		SchedulerHelper.cancelSchedule(scheduleHandler)
	end
	scheduleHandler = nil
end

function HeroVoiceManager:startPlayMainMenuVoice()
	local heroIds = G_UserData:getTeam():getHeroBaseIdsInBattle()

	local function getInterval()
		return math.random(15, 20)
	end

	local function playHeroVoice()
		if self._isInMainMenu then
    		local index = math.random(#heroIds)
    		local heroId = heroIds[index]
			self:playVoiceWithHeroId(heroId)
    	end
	end

	local function playLoop()
		local interval = getInterval()
		self._mainMenuScheduler = SchedulerHelper.newScheduleOnce(function()
			playHeroVoice()
			if self._isInMainMenu then
				playLoop()
			end
    	end, interval)
	end

	self:playCurRoleVoice()
	playLoop()
end

function HeroVoiceManager:stopPlayMainMenuVoice()
	if self._mainMenuScheduler then
		SchedulerHelper.cancelSchedule(self._mainMenuScheduler)
		self._mainMenuScheduler = nil
	end
end

return HeroVoiceManager